align/pnr/placer_pythonic.py

```diff
@@ -0,0 +1,116 @@
+import logging
+
+from align.schema.hacks import VerilogJsonTop
+from align.pnr.checker import check_placement
+from align.pnr.placer_pythonic_sp import place_using_sequence_pairs
+
+
+def compute_topoorder(modules, concrete_name, key='abstract_template_name'):
+    found_modules, found_leaves = set(), set()
+    order = list()
+
+    def aux(cn):
+        if cn in modules:
+            found_modules.add(cn)
+            for instance in modules[cn]['instances']:
+                ctn = instance[key]
+                if ctn not in found_modules and ctn not in found_leaves:
+                    aux(ctn)
+            order.append(cn)
+        else:
+            found_leaves.add(cn)
+            order.append(cn)
+    aux(concrete_name)
+
+    return order, found_modules, found_leaves
+
+
+def trim_placement_data(placement_data, top_level):
+
+    modules = {module['concrete_name']: module for module in placement_data['modules']}
+
+    top_concrete_names = [module['concrete_name'] for module in placement_data['modules'] if module['abstract_name'] == top_level]
+    all_modules_leaves = list()
+    for concrete_name in top_concrete_names:
```

Comment:
If you want, you can use a set() (instead of list()) on line 33, and then use update (instead of extend) on lines 36 and 37.
Then you can remove line 39.